perf-regression/gpu-margo-p2p-bw.cu

```diff
@@ -0,0 +1,796 @@
+static hg_bulk_t      g_bulk_handle = HG_BULK_NULL;
+
+static char*	      c_buffer = NULL;	/* Pointer to buffer via cudaMalloc */
+int use_cuda_buf = 0; 			/* Using cudaMalloc buf */
+
+int main(int argc, char** argv)
+{
+    margo_instance_id      mid;
+    int                    nranks;
+    int                    my_mpi_rank;
+    ssg_group_id_t         gid;
+    char*                  gid_buffer;
+    size_t                 gid_buffer_size;
+    int                    gid_buffer_size_int;
+    int                    namelen;
+    char                   processor_name[MPI_MAX_PROCESSOR_NAME];
+    int                    i;
+    ABT_xstream*           bw_worker_xstreams = NULL;
+    ABT_sched*             bw_worker_scheds   = NULL;
+    struct margo_init_info mii;
+    struct hg_init_info    hii;
+    int                    group_size;
+    int                    ret;
+    struct hg_bulk_attr    bulk_attr;
+
+    MPI_Init(&argc, &argv);
+
+    /* 2 process one-way bandwidth measurement only */
+    MPI_Comm_size(MPI_COMM_WORLD, &nranks);
+    if (nranks != 2) {
+        usage();
+        exit(EXIT_FAILURE);
+    }
+    MPI_Comm_rank(MPI_COMM_WORLD, &my_mpi_rank);
+    MPI_Get_processor_name(processor_name, &namelen);
+
+    ret = parse_args(argc, argv, &g_opts);
+    if (ret < 0) {
+        if (my_mpi_rank == 0) usage();
+        exit(EXIT_FAILURE);
+    }
+
+    /* allocate one big buffer for rdma transfers */
+    /* On server side, optionally use an mmap'd buffer.  Always calloc on
+     * client. */
+    if (my_mpi_rank == 0 && g_opts.mmap_filename) {
+        g_buffer = (char *)custom_mmap_alloc(g_opts.mmap_filename, g_opts.g_buffer_size,
+                                     my_mpi_rank);
+    } else {
+        g_buffer = NULL;
+        if (g_opts.align_buffer) {
+            ret = posix_memalign((void**)(&g_buffer), 4096,
+                                 g_opts.g_buffer_size);
+            if (ret != 0)
+                fprintf(stderr, "Error in posix_memalign: %s\n", strerror(ret));
+        } else
+            g_buffer = (char *)calloc(g_opts.g_buffer_size, 1);
+    }
+
+    if (!g_buffer) {
+        fprintf(stderr, "Error: unable to allocate %lu byte buffer.\n",
+                g_opts.g_buffer_size);
+        return (-1);
+    }
+
+    /* Shouldn't have -j and -k options at the same time */
+    if(g_opts.gpu_to_mem && g_opts.mem_to_gpu) {
+        fprintf(stderr, "Error: Invalid to have both -j and -k options simultaneously\n");
+        exit(EXIT_FAILURE);
+    }
+
+    /* Allocate GPU memory if (GPU to GPU, -j option for rank 0, -k option for rank 1 */
+    if ((!g_opts.gpu_to_mem && !g_opts.mem_to_gpu) ||
+       (my_mpi_rank == 0 && g_opts.gpu_to_mem) || 
+       (my_mpi_rank == 1 && g_opts.mem_to_gpu)) {
+    	cudaMalloc((void **)&c_buffer, g_opts.g_buffer_size);
+    	cudaMemset(c_buffer, 0, g_opts.g_buffer_size);
+    	cudaMemcpy(c_buffer, g_buffer, g_opts.g_buffer_size, cudaMemcpyHostToDevice);
```

Comment:
Is this cudaMemcpy needed? There shouldn't be anything in g_buffer at this point.

Reply:
Fixed